packages/in_app_purchase/lib/src/store_kit_wrappers/sk_product_request_wrapper.dart

```diff
@@ -0,0 +1,139 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/foundation.dart';
+import '../channel.dart';
+
+/// [SKProductRequestWrapper] wraps IOS SKProductRequest class to to retrive StoreKit product information in dart.
+///
+/// https://developer.apple.com/documentation/storekit/skproductsrequest?language=objc
+class SKProductRequestWrapper {
```

Comment:
I don't see much resemblance to SKProductRequestWrapper, SKRequest seems like something that represents a single App Store request and manages its state(start, cancel, callbacks).
While this a wrapper for a static method...

I'm not sure whether the bigger part of the method channel API we will end up building will be similar to what we do here, or to what we did with `canMakePayments`(basically a mirror of the App Store API).
If the former is the case, what do you guys think about having a (AppStore/AppStoreClient/FlutterAppStoreClient?) class which implements the Dart side of the method channel API?

@mklim

Reply:
I'm not totally sure I'm following. @amirh are you saying that this API is too different from `SkProductsRequest`? If so I agree here. Ideally I'd like everything in the `wrappers/` layer to be a thin wrapper around the original API, with as little a translation as possible but taking advantage of Dart's language features when appropriate (mostly translating callbacks to futures). I think for now it's a problem that the naming on this doesn't match up, and it's not really wrapping the response. This method should be returning the data from [`SkProductsRequestDelegate`](https://developer.apple.com/documentation/storekit/skproductsrequestdelegate?language=objc), so the [products and a list of invalid identifiers](https://developer.apple.com/documentation/storekit/skproductsresponse?language=objc). Right now it's missing information from the App Store API.

I mentioned in an earlier comment that this should ideally also be returning a cancelable instance but didn't want to block this PR on it. What I'm thinking of would still be a static method on the class but would be returning an instance of a cancelable request, which I think is a close enough match. The important thing is to make sure we're not dropping any functionality.

```
class SKProductRequestWrapper {
   static CancelableOperation<Future<SkProductsResponseWrapper>> initWithProductIdentifiers(List<String> productIdentifiers);
}

class SkProductsResponseWrapper {
  final List<SkProductsWrapper> products;
  final List<String> invalidIdentifiers;
}
```

Then it's more the job of the generic API that's a level up from here to adapt that and just return a single `List<Product>`.

>what do you guys think about having a (AppStore/AppStoreClient/FlutterAppStoreClient?) class which implements the Dart side of the method channel API?

Also not totally sure what you're suggesting here, can you explain more?

Reply:
OK, we caught up offline. Summarizing the discussion below.

So the goal for the `wrappers/` here is to ultimately expose all the platform functionality, and make sense at a glance given a previous understanding of how the underlying platform works. That said in a specific case like `SkRequest` wrapping each type of request as it's own class with a single static method isn't really worth it. It doesn't mirror the underlying platform enough to be instantly recognizable: `start`, `stop`, `cancel`, and the `delegate` all need to be cleared out to make sense in Dart. And the end result is really hard to read Dart, too.

So for something like this we should make a more idiomatic API, as long as the function is still totally understandable in terms of the underlying platform.

So instead of my original suggestion above,  collecting all the `SkRequest` methods into one place without redundant named classes for any of them. But clear naming for each of the methods so it's obvious what they are at a glance (or grep), and the returned data types still covering everything.

```
class SkRequester {
  Future<SkProductsResponseWrapper> startSkProductsRequest(List<String> productIdentifiers); // optionally CancelableOperation<Future<>> instead

  // etc
}

class SkProductsResponseWrapper {
  final List<SkProductsWrapper> products;
  final List<String> invalidIdentifiers;
}
```